### PR TITLE
Implement cookie interceptor

### PR DESCRIPTION
**Pull request by rhyshort**
Implement cookie interceptor, which logs a user in,
using the _session endpoint.

Tests use NSURLProtocol to mock out the responses
that would be received from the server.

Additionally change the `CDTHTTPInterceptorContext` to have a `NSHTTPURLResponse` object rather than a `NSURLResponse` Object.

## Reviewers

reviewer @tomblench 
reviewer @mikerhodes

### CODE REVIEW

**Comment by mikerhodes** [comment]
@rhyshort you need to remove the NSURLProtocol subclasses after e5cae0750b150f5da3c51c02ad117210f73ba8e4

**Comment by rhyshort** [comment]
@mikerhodes should have got them in 7e992ae

**Comment by mikerhodes** [comment]
@rhyshort you need to add copyrights to your new headers, i.e., these: https://github.com/cloudant/CDTDatastore/pull/200/files#diff-2bfcbb24be02e29952e2dcfb7e61d74cL6

File: Tests/Tests/CDTSessionCookieInterceptorTests.m [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+- (void)tearDown { [OHHTTPStubs removeAllStubs]; }
+- (void)testCookieInterceptorSuccessfullyGetsCookie
+{
+    NSString *expectedCookieString = @"AuthSession=cm9vdDo1MEJCRkYwMjq0LO0ylOIwShrgt8y-UkhI-c6BGw";
+    CDTSessionCookieInterceptor *interceptor =
+        [[CDTSessionCookieInterceptor alloc] initWithUsername:@"username" password:@"password"];
+
+    // override the NSURLSession so we can stub the responses
+    //    NSURLSessionConfiguration *config = [NSURLSessionConfiguration
```
Comment (tomblench) [inline Tests/Tests/CDTSessionCookieInterceptorTests.m]: commented out code?

File: Tests/Tests/CDTReplicationTests.m [hunk 2]
```diff
@@ -65,15 +64,22 @@ @interface CDTReplicationTests : CloudantSyncTests
...
 @implementation CDTReplicationTests
 
+- (void)setUp { [super setUp]; }
+- (void)tearDown { [super tearDown]; }
 - (void)testFiltersWithChangesFeed
 {
+    [OHHTTPStubs stubRequestsPassingTest:^BOOL(NSURLRequest *__nonnull request) {
+      return YES;
+    }
+        withStubResponse:^OHHTTPStubsResponse *__nonnull(NSURLRequest *__nonnull request) {
```
Comment (tomblench) [inline Tests/Tests/CDTReplicationTests.m]: indenting

File: Tests/Tests/CDTSessionCookieInterceptorTests.m (hunk 1 above)
Comment (rhyshort) [inline Tests/Tests/CDTSessionCookieInterceptorTests.m]: fixed in a4f929b

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                    cookie = nil;
+                }
+            } else if ([httpResp statusCode] == 401) {
+                CDTLogError(CDTREPLICATION_LOG_CONTEXT, @"Credentials are incorrect, cookie "
+                                                        @"authentication will not be attempted "
+                                                        @"again by this interceptor object");
+                self.shouldMakeCookieRequest = NO;
+            } else if ([httpResp statusCode] / 100 == 5) {
+                CDTLogError(CDTREPLICATION_LOG_CONTEXT,
+                            @"Failed to get cookie from the server, response code was %ld cookie "
```
Comment (tomblench) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: Two sentences: "...response code was %ld. Cookie..."

File: Tests/Tests/CDTReplicationTests.m (hunk 2 above)
Comment (rhyshort) [inline Tests/Tests/CDTReplicationTests.m]: `clang-format` sets the indenting, and insists that it is like this

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    if (!port) {
+        if ([scheme isEqualToString:@"https"]) {
+            port = @(443);
+        } else {
+            port = @(80);
+        }
+    }
+
+    NSURL *sessionUrl = [NSURL
+        URLWithString:[NSString stringWithFormat:@"%@://%@:%@/_session", scheme, [url host], port]];
```
Comment (tomblench) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: Is there a better way of doing this than string concatenation, using the various NSURL utilities?

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 3 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: done in 5f21438

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 4 above)
Comment (mikerhodes) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: Good spot. Now we're iOS 7 + we can use NSURLComponents, http://nshipster.com/nsurl/.

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:sessionUrl];
+    request.HTTPBody = [self.cookieRequestBody dataUsingEncoding:NSUTF8StringEncoding];
+    [request setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];
```
Comment (mikerhodes) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: Perhaps set on the session config, then it's next to where you build the body in `init`

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        } else {
+            port = @(80);
+        }
+    }
+
+    NSURL *sessionUrl = [NSURL
+        URLWithString:[NSString stringWithFormat:@"%@://%@:%@/_session", scheme, [url host], port]];
+
+    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:sessionUrl];
+    request.HTTPBody = [self.cookieRequestBody dataUsingEncoding:NSUTF8StringEncoding];
```
Comment (mikerhodes) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: You might as well make the data in `init`.

Comment (mikerhodes) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: Or even the whole request?

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+    dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+    __block NSString *cookie = nil;
+    NSURLSessionDataTask *task = [self.urlSession
+        dataTaskWithRequest:request
+          completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+            // all parameters to block are nullable, annotations removed
+            // to provide compatbility with xcode 6.4 and xcode 7
+            NSHTTPURLResponse *httpResp = (NSHTTPURLResponse *)response;
+            // handle response being nil
```
Comment (mikerhodes) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: Still needed?

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+            if ([httpResp statusCode] / 100 == 2) {
+                // we have the cookie. maybe
+                if (data && [self hasSessionStarted:data]) {
+                    // get the cookie from the header
+                    NSString *cookieHeader = [[httpResp allHeaderFields] valueForKey:@"Set-Cookie"];
+
+                    cookie = [cookieHeader componentsSeparatedByString:@";"][0];
+
```
Comment (mikerhodes) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: Probably don't need new lines on 101 or 103

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+                } else {
+                    cookie = nil;
+                }
```
Comment (mikerhodes) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: No need for the else block?

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+                                                        @"authentication will not be attempted "
+                                                        @"again by this interceptor object");
+                self.shouldMakeCookieRequest = NO;
+            } else if ([httpResp statusCode] / 100 == 5) {
+                CDTLogError(CDTREPLICATION_LOG_CONTEXT,
+                            @"Failed to get cookie from the server, response code was %ld cookie "
+                            @"authentication will not be attempted again by this interceptor "
+                            @"object",
+                            (long)[httpResp statusCode]);
+                self.shouldMakeCookieRequest = NO;
```
Comment (mikerhodes) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: I don't think this is valid for `500` status codes. `500` errors tend to be transient.

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+            } else {
+                CDTLogError(CDTREPLICATION_LOG_CONTEXT,
+                            @"Failed to get cookie from the server,response code %ld cookie "
+                            @"authentication will not be attempted again by this interceptor "
+                            @"object",
+                            (long)[httpResp statusCode]);
```
Comment (mikerhodes) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: `will not be attempted again by this interceptor` is not true in this block, I think.

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+            }
+
+            dispatch_semaphore_signal(sema);
+
+          }];
+    [task resume];
+    dispatch_semaphore_wait(sema, dispatch_time(DISPATCH_TIME_NOW, 600 * NSEC_PER_SEC));
```
Comment (mikerhodes) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: Probably want to break out the 600 I stuck here into a constant.

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+- (BOOL)hasSessionStarted:(nonnull NSData *)data
+{
+    NSDictionary *jsonResponse = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
+
+    // only check for ok:true, https://issues.apache.org/jira/browse/COUCHDB-1356
+    // means we cannot check that the name returned is the one we sent.
+    return
+        [jsonResponse objectForKey:@"ok"] != nil && [[jsonResponse objectForKey:@"ok"] boolValue];
```
Comment (mikerhodes) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: You can use `jsonResponse[@"ok"]`. If you shorted the var name to `json` you'll be able to fit everything on one line.

In particular, I think `[nil boolValue]` is `NO` -- double check, but you might not need the guarding clause.

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 4 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: Yes `NSURLComponents`, changed to using it in be47e06

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+
+@implementation CDTSessionCookieInterceptor
+
+- (instancetype)initWithUsername:(NSString *)username password:(NSString *)password
+{
+    self = [super init];
+    if (self) {
+        _cookieRequestBody = [NSString stringWithFormat:@"name=%@&password=%@", username, password];
+        _shouldMakeCookieRequest = YES;
+        NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];
```
Comment (mikerhodes) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: I'd make this an ephemeral session, given there's no need for caching or anything much really.

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 7 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: Yes. We need to make the cast to make it easier to work with the http information we need from it

File: Tests/Tests/CDTSessionCookieInterceptorTests.m [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+                                     @"AuthSession=cm9vdDo1MEJCRkYwMjq0LO0ylOIwShrgt8y-UkhI-c6BGw; "
+                                     @"Version=1; Path=/; HttpOnly"
+                             }];
+        }];
+}
+
+- (void)tearDown { [OHHTTPStubs removeAllStubs]; }
+- (void)testCookieInterceptorSuccessfullyGetsCookie
+{
+    NSString *expectedCookieString = @"AuthSession=cm9vdDo1MEJCRkYwMjq0LO0ylOIwShrgt8y-UkhI-c6BGw";
```
Comment (mikerhodes) [inline Tests/Tests/CDTSessionCookieInterceptorTests.m]: You want to make this a constant so you can share it between the response and the expection?

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 7 above)
Comment (mikerhodes) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: I meant the comment -- it didn't seem to be addressed in the `if` statement, I was wondering whether it was out of date.

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 13 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: quick test with LLDB shows this:
```
(lldb) expr [(NSNumber*)nil boolValue]
(BOOL) $2 = NO
```

So I'll remove the guard

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 9 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: removed in 09d5302

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 8 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: removed in 09d5302

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 6 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: We can't make the request ahead of time without knowing the URL of the couchdb instance.

Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: But I can move the header setting into to init, and add it to the session instead

Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: see 5b180ee

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 5 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: done in 5b180ee

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 13 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: done in 3274138

File: Tests/Tests/CDTSessionCookieInterceptorTests.m (hunk 15 above)
Comment (rhyshort) [inline Tests/Tests/CDTSessionCookieInterceptorTests.m]: done in a02b887

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 13 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: done in  3274138

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 12 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: done in  6740dd0

**Comment by tomblench** [comment]
+1

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 11 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: fixed in 9f47b0a

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 10 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: done in b0f5798

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 7 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: handled `nil` in 56bc3da

File: Classes/common/HTTP/CDTSessionCookieInterceptor.m (hunk 14 above)
Comment (rhyshort) [inline Classes/common/HTTP/CDTSessionCookieInterceptor.m]: done in e3c1c13

**Comment by mikerhodes** [comment]
+1 I think we've got the functionality right -- squash time.

**Pull request merged**